### PR TITLE
[DoNotMerge][WIP] add support for thumb* (Cortex-M) targets

### PR DESCRIPTION
**Pull request by japaric**
**NOTE** Do not merge. rust-lang/rfcs#1645 must be approved first.

As specified in the ("less targets" alternative of the) RFC, this PR adds these new, baseline
targets to the compiler: 
- `thumbv6m-none-eabi`: Cortex-M0, Cortex-M0+, Cortex-M1
- `thumbv7m-none-eabi`: Cortex-M3
- `thumbv7em-none-eabi`: Cortex-M4, Cortex-M7 -- no FPU, soft float ABI
- `thumbv7em-none-eabihf`: Cortex-M4, Cortex-M7 -- SP FPU, hard float ABI

And tweaks the build script of compiler_builtins to support these targets as well.

With these changes it's possible to compile the following subsets of standard crates for each of
these targets:
- For `thumbv6m-none-eabi`: `compiler_builtins`, `core` and `rustc_unicode`. (*)
- For the other 3 targets: `collections`, `compiler_builtins`, `rand` and every other crate below
  them.

(*) This target doesn't support atomic operations so crates like `alloc`, which depend on atomics
(e.g. `Arc`), and crates above them can't be compiled for this target.

TODOs
- [ ] The corresponding RFC must be approved.
- [x] Teach rustbuild to not build the whole set of standard crates for these targets. Instead tell
  it that it must compile only the smaller set of crates that each target supports.
- [ ] This depends on rust-lang/rust#36512. That must land first.

cc @alexcrichton @brson

### CODE REVIEW

File: src/libcompiler_builtins/build.rs [hunk 1]
```diff
@@ -90,6 +113,25 @@ fn main() {
...
+    if llvm_target[0].starts_with("thumb") {
+        cfg.flag("-mthumb");
+    }
+
+    if llvm_target[0] == "thumbv6m" {
+        cfg.flag("-march=armv6-m");
+    }
+
+    if llvm_target[0] == "thumbv7m" {
+        cfg.flag("-march=armv7-m");
```
Comment (alexcrichton) [inline src/libcompiler_builtins/build.rs]: This logic should end up in gcc-rs, right?

File: src/libcompiler_builtins/Cargo.toml [hunk 2]
```diff
@@ -13,3 +13,4 @@ core = { path = "../libcore" }
 
 [build-dependencies]
 gcc = "0.3.27"
+rustc-cfg = { git = "https://github.com/japaric/rustc-cfg", branch = "llvm-target" }
```
Comment (alexcrichton) [inline src/libcompiler_builtins/Cargo.toml]: It looks like this may be something that should be in gcc-rs, but maybe I'm missing something? Couldn't we just look at `$TARGET` like we do for other platforms?

File: src/librustc_back/target/mod.rs [hunk 3]
```diff
@@ -63,11 +63,12 @@ mod netbsd_base;
 mod solaris_base;
 mod windows_base;
 mod windows_msvc_base;
+mod thumb_base;
 
 pub type TargetResult = Result<Target, String>;
 
 macro_rules! supported_targets {
-    ( $(($triple:expr, $module:ident)),+ ) => (
+    ( $(($triple:expr, $module:ident)),+, ) => (
```
Comment (alexcrichton) [inline src/librustc_back/target/mod.rs]: You can actually move the comma inside the parens to require a trailing comma like:

```rust
( $(($triple:expr, $module:ident),)+ )
```

File: src/librustc_back/target/mod.rs [hunk 4]
```diff
@@ -391,7 +397,13 @@ impl Default for TargetOptions {
...
             has_elf_tls: false,
             obj_is_bitcode: false,
-            max_atomic_width: 0,
+            // NOTE this is *not* the real default value. The default value of max_atomic_width is
+            // actually the pointer width of the target. This default is injected in the
+            // Target::from_json function. Still, we have to pick some value to put here. We'll pick
+            // an impossible value: u64::max_value() because using a valid value like 0 or 8 as the
+            // default would cause the max-atomic-width field to be "lost" (it would get replaced
+            // by target_pointer_width) during the Target <-> JSON round trip
+            max_atomic_width: u64::max_value(),
```
Comment (alexcrichton) [inline src/librustc_back/target/mod.rs]: Hm I'm not sure I quite understand why this is necessary, could you elaborate what you mean by "lost" in the comment?

File: src/librustc_back/target/thumbv6m_none_eabi.rs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        target_pointer_width: "32".to_string(),
+        data_layout: "e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64".to_string(),
+        arch: "arm".to_string(),
+        target_os: "none".to_string(),
+        target_env: "".to_string(),
+        target_vendor: "".to_string(),
+
+        options: TargetOptions {
+            // NOTE prevents mis-optimizations of `ptr::copy_nonoverlapping` when unaligned loads
+            // are involved
```
Comment (alexcrichton) [inline src/librustc_back/target/thumbv6m_none_eabi.rs]: Is this a FIXME or just a NOTE? (e.g. is it a bug we should fix at some point?)

File: src/rustc/std_shim/Cargo.toml [hunk 6]
```diff
@@ -40,11 +40,23 @@ debug = false
...
 # Reexport features from std
 [features]
-jemalloc = ["std/jemalloc"]
-debug-jemalloc = ["std/debug-jemalloc"]
 backtrace = ["std/backtrace"]
+debug-jemalloc = ["std/debug-jemalloc"]
+jemalloc = ["std/jemalloc"]
+full = ["core", "std"]
+thumb_no_atomics  = ["compiler_builtins", "core", "rustc_unicode"]
+thumb = ["core", "compiler_builtins", "collections", "rand"]
```
Comment (alexcrichton) [inline src/rustc/std_shim/Cargo.toml]: Perhaps these could be give non-thumb specific names and just mention `no-atomics` and `atomics`?

**Review by alexcrichton — COMMENTED** [review]
Awesome, thanks @japaric! Could you also add comments to all the new target files indicating what the targets are meant to be used for as well? (e.g. justify target-features and whatnot)

File: src/librustc_back/target/mod.rs (hunk 4 above)
Comment (japaric) [inline src/librustc_back/target/mod.rs]: Sure! Let me ellaborate here, first, what happens today without this change:

- All targets go through a round trip from `TargetOptions` to JSON and back to `TargetOptions`.
- Let's say you set `max_atomic_width` to the `Default` value, which is `0` today, for a target specification, when this specification is converted to JSON the `max_atomic_width` field is *not* present in that JSON representation.
- When that JSON representation is converted back to `TargetOptions`, the current logic sets `max_atomic_width` to `target_pointer_witdh`, i.e. `32` or `64`, because the `max_atomic_width` field is not present in the JSON representation.

TL;DR You set `max_atomic_witdth` to `0` but the compiler is actually using `32` or `64` for the target.

The bug is subtle and the proposed fix is sort of magical. Perhaps we could solve it in a better way.

File: src/libcompiler_builtins/Cargo.toml (hunk 2 above)
Comment (japaric) [inline src/libcompiler_builtins/Cargo.toml]: This change is from rust-lang/rust#36512. Check that PR for context.

**Comment by japaric** [comment]
@alexcrichton I don't know if you have have seen it yet. But the tools is considering only adding the target specifications and not building binary artifacts cf. https://github.com/rust-lang/rfcs/pull/1645#issuecomment-248463998.

I'll come back to this once that's decided.

**Comment by alexcrichton** [comment]
Sounds good to me!

File: src/librustc_back/target/thumbv6m_none_eabi.rs (hunk 5 above)
Comment (japaric) [inline src/librustc_back/target/thumbv6m_none_eabi.rs]: After actually checking the documentation. The ARMv6-M Architecture Reference Manual says:

> A3.2 Alignment Support
> ARMv6-M always generates a fault when an unaligned access occurs.

So no, this can't be "fixed". The architecture doesn't support this operation so it's correct to disable it here.

**Comment by japaric** [comment]
closing in favor of #36874 which only adds the target definitions
